Workflow/Sources/WorkflowObserver.swift

```diff
@@ -0,0 +1,322 @@
+    /// - Returns: An optional closure to be called immediately after the new `Rendering` is produced, which takes the rendering as the only parameter.
+    func workflowWillRender<WorkflowType: Workflow>(
+        _ workflow: WorkflowType,
+        state: WorkflowType.State,
+        session: WorkflowSession
+    ) -> ((WorkflowType.Rendering) -> Void)?
+
+    /// Called after an existing `Workflow` is updated.
+    /// - Parameters:
+    ///   - oldWorkflow: The previous `Workflow`
+    ///   - newWorkflow: The new `Workflow`
+    ///   - state: The state **after** the update has occurred.
+    ///   - session: The `WorkflowSession` corresponding to the backing `WorkflowNode`.
+    func workflowDidChange<WorkflowType: Workflow>(
+        from oldWorkflow: WorkflowType,
+        to newWorkflow: WorkflowType,
+        state: WorkflowType.State,
+        session: WorkflowSession
+    )
+
+    /// Called after a `WorkflowAction` is received, but **before** it has been handled and propagated up the tree.
+    /// - Parameters:
+    ///   - action: The action that was received.
+    ///   - session: The `WorkflowSession` corresponding to the backing `WorkflowNode`.
+    func workflowDidReceiveAction<Action: WorkflowAction>(
+        _ action: Action,
+        workflow: Action.WorkflowType,
+        session: WorkflowSession
+    )
+
+    /// Called when a `WorkflowAction` will be applied to its corresponding Workflow's State
+    /// - Parameters:
+    ///   - action: The action that will be applied.
+    ///   - workflow: The action's corresponding `Workflow`.
+    ///   - state: The state to which the action will be applied.
+    ///   - session: The `WorkflowSession` corresponding to the backing `WorkflowNode`.
+    /// - Returns: An optional closure to be called immediately after the action is applied to the `State`, and an optional `Output` has been produced. The closure takes the updated state and optional output as its arguments.
+    func workflowWillApplyAction<Action: WorkflowAction>(
+        _ action: Action,
+        workflow: Action.WorkflowType,
+        state: Action.WorkflowType.State,
+        session: WorkflowSession
+    ) -> ((Action.WorkflowType.State, Action.WorkflowType.Output?) -> Void)?
+}
+
+// MARK: - WorkflowSession
+
+/// A `WorkflowSession`encapsulates the information that gives a `WorkflowNode` its identity.
+/// In particular, it captures:
+/// - The type of the corresponding `Workflow` conformance.
+/// - The `String` key used when the workflow was rendered.
+/// - An `Identifier` type that is unique across program execution.
+/// - An optional reference to a parent `WorkflowSession`, to differentiate root nodes.
+public struct WorkflowSession {
+    public struct Identifier: Hashable {
+        private static var _nextRawID: UInt64 = 0
+        private static func _makeNextSessionID() -> UInt64 {
+            _nextRawID += 1
+            return _nextRawID
+        }
```

Comment:
We might want to mention in the docstring that `WorkflowSession.init()` is not thread-safe.